cmd/deployment-check/deployment.go

```diff
@@ -0,0 +1,703 @@
+				if !ok {
+					log.Infoln("Got a watch event for a non-deployment object -- ignoring.")
+					continue
+				}
+
+				log.Debugln("Received an event watching for deployment changes:", d.Name, "got event", event.Type)
+
+				// We want an event type of DELETED here.
+				if event.Type == "DELETED" {
+					log.Infoln("Received a", event.Type, "while watching for deployment with name ["+d.Name+"] to be deleted.")
+					deleteChan <- nil
+					return
+				}
+			}
+
+			watch.Stop()
+		}
+	}()
+
+	log.Infoln("Attempting to delete deployment in", checkNamespace, "namespace.")
+	if client == nil {
+		err := errors.New("nil kubernetes client")
+		return err
+	}
+
+	// Make a delete options object to delete the deployment.
+	deletePolicy := metav1.DeletePropagationForeground
+	graceSeconds := int64(1)
+	deleteOpts := metav1.DeleteOptions{
+		GracePeriodSeconds: &graceSeconds,
+		PropagationPolicy:  &deletePolicy,
+	}
+
+	// Delete the deployment.
+	err := client.AppsV1().Deployments(checkNamespace).Delete(checkDeploymentName, &deleteOpts)
+	if err != nil {
+		log.Infoln("Could not delete deployment:", checkDeploymentName)
+		log.Infoln("Beginning backoff retry loop to delete deployment.")
+	}
+
+	return <-deleteChan
+}
+
+// cleanUpOrphanedDeployment cleans up deployments created from previous checks.
+func cleanUpOrphanedDeployment() error {
+
+	cleanUpChan := make(chan error)
+
+	go func() {
+		defer close(cleanUpChan)
+
+		// Watch that it is gone.
+		watch, err := client.AppsV1().Deployments(checkNamespace).Watch(metav1.ListOptions{
+			Watch:         true,
+			FieldSelector: "metadata.name=" + checkDeploymentName,
+			// LabelSelector: defaultLabelKey + "=" + defaultLabelValueBase + strconv.Itoa(int(now.Unix())),
+		})
+		if err != nil {
+			log.Infoln("Error creating watch client.")
+			cleanUpChan <- err
+			return
+		}
+
+		// Stop the watch on each loop because we will create a new one.
+		defer watch.Stop()
+
+		for {
+
+			// Watch for a DELETED event.
+			select {
+			case event := <-watch.ResultChan():
+
+				d, ok := event.Object.(*v1.Deployment)
+				if !ok {
+					log.Infoln("Got a watch event for a non-deployment object -- ignoring.")
+					continue
+				}
+
+				log.Debugln("Received an event watching for deployment changes:", d.Name, "got event", event.Type)
+
+				// We want an event type of DELETED here.
+				if event.Type == "DELETED" {
+					log.Infoln("Received a", event.Type, "while watching for deployment with name ["+d.Name+"] to be deleted")
+					cleanUpChan <- nil
+					return
+				}
+			case done := <-waitForDeploymentToDelete():
+				log.Infoln("Received a complete while waiting for service to delete:", done)
+				cleanUpChan <- nil
+				return
+			case <-ctx.Done():
+				log.Errorln("context expired while waiting for service to be cleaned up.")
+				cleanUpChan <- nil
+				return
+			}
+		}
+	}()
+
+	log.Infoln("Removing previous deployment in", checkNamespace, "namespace.")
+	if client == nil {
```

Comment:
check this upstream rather than in all downstream usages

Reply:
removed